# Putting It Together 2: Answering Questions By Exploring and Plotting the Dataset (Exercise 1)

Dataset: One waiter recorded information about each tip he received over a period of a few months working in one restaurant.

Columns:
- total_bill (\$)
- tip (\$)
- sex (Male, Female)
- smoker (Yes, No)
- day (Thurs, Fri, Sat, Sun)
- time (Lunch, Dinner)
- size (# of people in party)

## Question 1. &nbsp;&nbsp; How does the tip left change with respect to each variable?

#!action

Q1. Which variable(s) is/are dependent?

#!response

#!action

Q2. Which variable(s) is/are independent?

#!response

#!action

Read in the dataset `datasets/tips_original.csv` and display its contents.

### Step 1. &nbsp;&nbsp;&nbsp; Read CSV Data into Pandas Dataframe

#### Substep. &nbsp;&nbsp; Import Pandas Library

#!blhint

- Create variable &nbsp;`pd`
- `import pandas as pd`

<details><summary>Blockly Hints</summary><ul><li>Blocks<ul><li><code>import .. as ..</code><ul><li>The <code>import .. as ..</code> block is found under IMPORT.</li> <li>A variable must have been created before it will appear in the dropdown.</li></ul></li></ul></li></ul></details>

In [1]:
#!response

import pandas as pd

#### Substep. &nbsp;&nbsp; Read CSV data and Save in Variable

#!blhint

- Create variable &nbsp;`df`
- `set df to` &nbsp;:&nbsp; `with pd do read_csv using` &nbsp;<-&nbsp; `" datasets/tips_original.csv "`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li><li><code>A</code> &nbsp;&lt;-&nbsp; <code>B</code>&nbsp; means insert block B into the hole in block A.</li></ul></li><li>Blocks<ul><li><code>set .. to</code><ul><li>The <code>set .. to</code> block is found under VARIABLES.</li> <li>A variable must have been created before it will appear in the dropdown.</li></ul></li><li><code>with .. do .. using ..</code><ul><li>The <code>with .. do .. using ..</code> block is found under VARIABLES.</li> <li>If <code>do</code> dropdown says "!Not populated until you execute code", click anywhere in the notebook tab, then click try "Run All Above Selected Cell" from the "Run" menu.</li> <li>If <code>with .. do .. using ..</code> block does not want to snap together nicely with the <code>set .. to</code> block, try dragging the <code>set .. to</code> block instead.</li></ul></li><li><code>" .. "</code><ul><li>The <code>" .. "</code> block is found under TEXT.</li></ul></li></ul></li></ul></details>

In [2]:
#!response

df = pd.read_csv('datasets/tips_original.csv')

#### Substep. &nbsp;&nbsp; Display Dataframe Contents

#!blhint

- variable &nbsp;`df`

<details><summary>Blockly Hints</summary><ul><li>Blocks<ul><li>variable<ul><li>After it is created, each variable has its own block at the end of the VARIABLES menu.</li></ul></li></ul></li></ul></details>

In [3]:
#!response

df

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

#!action

Look at the contents of the dataframe to answer the following questions.

#!action

Q1. Which variables in the dataset are numeric?

#!response

#!action

Q2. Which variables in the dataset are categorical?

#!response

In this dataset, the value of tip may depend too heavily on the value of total_bill, since customers will generally leave a percentage of the total cost as a tip. If we calculate the tip percentage left and use that as the dependent variable, it will make it easier to determine how the tip left changes with the other variables like day, time, size, etc.

#!action

For each tip value, calculate the tip as a percentage of total_bill and add the results to the dataframe in a new column.

### Step 2. &nbsp;&nbsp;&nbsp; Aggregate Data in New Column

#!action

We can calculate the tip percentage with the formula:  

Tip Percent = Tip / Total Bill

#### Substep. &nbsp;&nbsp; Perform the Calculation and Store in Variable 

#!blhint

- Create variable &nbsp;`tip_percent`
- `set tip_percent to` &nbsp;:&nbsp; ` .. * .. ` &nbsp;<-
  - ` .. / .. ` &nbsp;<-
    - `from df get tip`
    - `from df get total_bill`
  - `100`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li><li><code>A</code> &nbsp;&lt;-&nbsp; <code>B</code>&nbsp; means insert block B into the hole in block A.</li></ul></li><li>Blocks<ul><li><code>set .. to</code><ul><li>The <code>set .. to</code> block is found under VARIABLES.</li> <li>A variable must have been created before it will appear in the dropdown.</li></ul></li><li><code>from .. get ..</code><ul><li>The <code>from .. get ..</code> block is found under VARIABLES.</li></ul></li></ul></li></ul></details>

In [4]:
#!response

tip_percent = df['tip'] / df['total_bill'] * 100

#### Substep. &nbsp;&nbsp; Insert the Contents of the Variable as a New Column

#!blhint

- Grab a freestyle block, type &nbsp;`df['tip_percent'] = tip_percent`

<details><summary>Blockly Hints</summary><ul><li>Blocks<ul><li>freestyle<ul><li>Unless specifically instructed, use the first block from the FREESTYLE menu.</li></ul></li></ul></li></ul></details>

In [5]:
#!response

df['tip_percent'] = tip_percent

#### Substep. &nbsp;&nbsp; Display Dataframe Contents

#!blhint

- variable &nbsp;`df`

<details><summary>Blockly Hints</summary><ul><li>Blocks<ul><li>variable<ul><li>After it is created, each variable has its own block at the end of the VARIABLES menu.</li></ul></li></ul></li></ul></details>

In [6]:
#!response

df

total_bill   tip     sex smoker   day    time  size  tip_percent
0         16.99  1.01  Female     No   Sun  Dinner     2     5.944673
1         10.34  1.66    Male     No   Sun  Dinner     3    16.054159
2         21.01  3.50    Male     No   Sun  Dinner     3    16.658734
3         23.68  3.31    Male     No   Sun  Dinner     2    13.978041
4         24.59  3.61  Female     No   Sun  Dinner     4    14.680765
..          ...   ...     ...    ...   ...     ...   ...          ...
239       29.03  5.92    Male     No   Sat  Dinner     3    20.392697
240       27.18  2.00  Female    Yes   Sat  Dinner     2     7.358352
241       22.67  2.00    Male    Yes   Sat  Dinner     2     8.822232
242       17.82  1.75    Male     No   Sat  Dinner     2     9.820426
243       18.78  3.00  Female     No  Thur  Dinner     2    15.974441

[244 rows x 8 columns]

### Step 3. &nbsp;&nbsp;&nbsp; Describe the Numerical Variables

#### Substep. &nbsp;&nbsp; Run describe()

#!blhint

- `with df do describe using`

<details><summary>Blockly Hints</summary><ul><li>Blocks<ul><li><code>with .. do .. using ..</code><ul><li>The <code>with .. do .. using ..</code> block is found under VARIABLES.</li> <li>If <code>do</code> dropdown says "!Not populated until you execute code", click anywhere in the notebook tab, then click try "Run All Above Selected Cell" from the "Run" menu.</li> <li>If <code>with .. do .. using ..</code> block does not want to snap together nicely with the <code>set .. to</code> block, try dragging the <code>set .. to</code> block instead.</li></ul></li></ul></li></ul></details>

In [7]:
#!response

df.describe()

total_bill         tip        size  tip_percent
count  244.000000  244.000000  244.000000   244.000000
mean    19.785943    2.998279    2.569672    16.080258
std      8.902412    1.383638    0.951100     6.107220
min      3.070000    1.000000    1.000000     3.563814
25%     13.347500    2.000000    2.000000    12.912736
50%     17.795000    2.900000    2.000000    15.476977
75%     24.127500    3.562500    3.000000    19.147549
max     50.810000   10.000000    6.000000    71.034483

#!action

Look at the output of describe() to answer the following questions.

#!action

Q1. What is the average tip percent?

#!response

### Step 4. &nbsp;&nbsp;&nbsp; Describe the Categorical Variables

#### Substep. &nbsp;&nbsp; Create a Subset of the Dataframe

#!blhint

- Create variable &nbsp;`categorical`
- `set categorical to` &nbsp;:&nbsp; `df [ .. ]` &nbsp;<-&nbsp; `create list with` &nbsp;:
  - `" sex "`
  - `" smoker "`
  - `" day "`
  - `" time "`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li><li><code>A</code> &nbsp;&lt;-&nbsp; <code>B</code>&nbsp; means insert block B into the hole in block A.</li></ul></li><li>Blocks<ul><li><code>set .. to</code><ul><li>The <code>set .. to</code> block is found under VARIABLES.</li> <li>A variable must have been created before it will appear in the dropdown.</li></ul></li><li><code>\{dictVariable\} [ .. ]</code><ul><li>The <code>\{dictVariable\} [ .. ]</code> block is found under LISTS.</li></ul></li><li><code>create list with</code><ul><li>The <code>create list with</code> block is found under LISTS.</li> <li>You can use the <code>+ -</code> controls on the block to change the number of notches. Unless specifically instructed, the block should not have any empty notches when you click Blocks to Code. </li></ul></li><li><code>" .. "</code><ul><li>The <code>" .. "</code> block is found under TEXT.</li></ul></li></ul></li></ul></details>

In [8]:
#!response

categorical = df[['sex', 'smoker', 'day', 'time']]

#### Substep. &nbsp;&nbsp; Run describe()

#!blhint

- `with categorical do describe using`

<details><summary>Blockly Hints</summary><ul><li>Blocks<ul><li><code>with .. do .. using ..</code><ul><li>The <code>with .. do .. using ..</code> block is found under VARIABLES.</li> <li>If <code>do</code> dropdown says "!Not populated until you execute code", click anywhere in the notebook tab, then click try "Run All Above Selected Cell" from the "Run" menu.</li> <li>If <code>with .. do .. using ..</code> block does not want to snap together nicely with the <code>set .. to</code> block, try dragging the <code>set .. to</code> block instead.</li></ul></li></ul></li></ul></details>

In [9]:
#!response

categorical.describe()

sex smoker  day    time
count    244    244  244     244
unique     2      2    4       2
top     Male     No  Sat  Dinner
freq     157    151   87     176

#!action

Look at the output of describe() to answer the following questions.

#!action

Q1. What is the mode value of each variable and how many times did it appear? (Hint: look at the rows `top` and `freq`)

#!response

#!action

Q2. If `Male` is the top value and it appears 157 times, how many times does `Female` appear? (Hint: count - freq)

#!response

#!action

Q3. How many smokers were there?

#!response

#!action

Q4. How many parties came for lunch?

#!response

`describe()` gives the summary stats for each column in the dataframe without respect to any other column. Therefore, if we want to figure out the average tip percent only when the party came during lunch, we have to either create a subset of the dataframe and run describe on the subset, or we can use box plots.

#!action

Generate box plots for the following:

- Tip Percent vs Sex
- Tip Percent vs Smoker
- Tip Percent vs Day
- Tip Percent vs Time

### Step 5. &nbsp;&nbsp;&nbsp; Generate Plotly Chart(s)

#### Substep. &nbsp;&nbsp; Import Plotly Express Library


#!blhint

- Create variable &nbsp;`px`
- `import plotly.express as px`

<details><summary>Blockly Hints</summary><ul><li>Blocks<ul><li><code>import .. as ..</code><ul><li>The <code>import .. as ..</code> block is found under IMPORT.</li> <li>A variable must have been created before it will appear in the dropdown.</li></ul></li></ul></li></ul></details>

In [10]:
#!response

import plotly.express as px

#!action

Generate the box plot for Tip Percent vs Sex.

#### Substep. &nbsp;&nbsp; Set Columns as Category and Value


#!blhint

- Create variable &nbsp;`category`
- `set category to` &nbsp;:&nbsp; `" sex "`
- Create variable &nbsp;`values`
- `set values to` &nbsp;:&nbsp; `" tip_percent "`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li></ul></li><li>Blocks<ul><li><code>set .. to</code><ul><li>The <code>set .. to</code> block is found under VARIABLES.</li> <li>A variable must have been created before it will appear in the dropdown.</li></ul></li><li><code>" .. "</code><ul><li>The <code>" .. "</code> block is found under TEXT.</li></ul></li></ul></li></ul></details>

In [11]:
#!response

category = 'sex'
values = 'tip_percent'

#### Substep. &nbsp;&nbsp; Generate Box Plot

#!blhint

- Grab a freestyle block, type &nbsp;`x=category`
- Grab a freestyle block, type &nbsp;`y=values`
- `with px do box using` &nbsp;<-&nbsp; `create list with` &nbsp;:
    - variable `df`
    - `x=category`
    - `y=values`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li><li><code>A</code> &nbsp;&lt;-&nbsp; <code>B</code>&nbsp; means insert block B into the hole in block A.</li></ul></li><li>Blocks<ul><li>freestyle<ul><li>Unless specifically instructed, use the first block from the FREESTYLE menu.</li></ul></li><li><code>with .. do .. using ..</code><ul><li>The <code>with .. do .. using ..</code> block is found under VARIABLES.</li> <li>If <code>do</code> dropdown says "!Not populated until you execute code", click anywhere in the notebook tab, then click try "Run All Above Selected Cell" from the "Run" menu.</li> <li>If <code>with .. do .. using ..</code> block does not want to snap together nicely with the <code>set .. to</code> block, try dragging the <code>set .. to</code> block instead.</li></ul></li><li><code>create list with</code><ul><li>The <code>create list with</code> block is found under LISTS.</li> <li>You can use the <code>+ -</code> controls on the block to change the number of notches. Unless specifically instructed, the block should not have any empty notches when you click Blocks to Code. </li></ul></li><li>variable<ul><li>After it is created, each variable has its own block at the end of the VARIABLES menu.</li></ul></li></ul></li></ul></details>

In [12]:
#!response

px.box(df, x=category, y=values)

#!action

Look at the chart to answer the following questions.

#!action

Q1. Compare the 5 number summary for the Male box plot to the Female box plot.

#!response

#!action

Q2. After seeing the box plots, do you think the tip percent depends on whether they were male or female?

#!response

#!action

Generate the box plot for Tip Percent vs Smoker.

#### Substep. &nbsp;&nbsp; Set Columns as Category and Value


#!blhint


- `set category to` &nbsp;:&nbsp; `" smoker "`
- `set values to` &nbsp;:&nbsp; `" tip_percent "`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li></ul></li><li>Blocks<ul><li><code>set .. to</code><ul><li>The <code>set .. to</code> block is found under VARIABLES.</li> <li>A variable must have been created before it will appear in the dropdown.</li></ul></li><li><code>" .. "</code><ul><li>The <code>" .. "</code> block is found under TEXT.</li></ul></li></ul></li></ul></details>

In [13]:
#!response

category = 'smoker'
values = 'tip_percent'

#### Substep. &nbsp;&nbsp; Generate Box Plot

#!blhint

- Grab a freestyle block, type &nbsp;`x=category`
- Grab a freestyle block, type &nbsp;`y=values`
- `with px do box using` &nbsp;<-&nbsp; `create list with` &nbsp;:
    - variable `df`
    - `x=category`
    - `y=values`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li><li><code>A</code> &nbsp;&lt;-&nbsp; <code>B</code>&nbsp; means insert block B into the hole in block A.</li></ul></li><li>Blocks<ul><li>freestyle<ul><li>Unless specifically instructed, use the first block from the FREESTYLE menu.</li></ul></li><li><code>with .. do .. using ..</code><ul><li>The <code>with .. do .. using ..</code> block is found under VARIABLES.</li> <li>If <code>do</code> dropdown says "!Not populated until you execute code", click anywhere in the notebook tab, then click try "Run All Above Selected Cell" from the "Run" menu.</li> <li>If <code>with .. do .. using ..</code> block does not want to snap together nicely with the <code>set .. to</code> block, try dragging the <code>set .. to</code> block instead.</li></ul></li><li><code>create list with</code><ul><li>The <code>create list with</code> block is found under LISTS.</li> <li>You can use the <code>+ -</code> controls on the block to change the number of notches. Unless specifically instructed, the block should not have any empty notches when you click Blocks to Code. </li></ul></li><li>variable<ul><li>After it is created, each variable has its own block at the end of the VARIABLES menu.</li></ul></li></ul></li></ul></details>

In [14]:
#!response

px.box(df, x=category, y=values)

#!action

Look at the chart to answer the following questions.

#!action

Q1. Compare the 5 number summary for the Yes box plot to the No box plot.

#!response

#!action

Q2. After seeing the box plots, do you think the tip percent depends on whether they were a smoker?

#!response

#!action

Generate the box plot for Tip Percent vs Day.

#### Substep. &nbsp;&nbsp; Set Columns as Category and Value


#!blhint

- `set category to` &nbsp;:&nbsp; `" day "`
- `set values to` &nbsp;:&nbsp; `" tip_percent "`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li></ul></li><li>Blocks<ul><li><code>set .. to</code><ul><li>The <code>set .. to</code> block is found under VARIABLES.</li> <li>A variable must have been created before it will appear in the dropdown.</li></ul></li><li><code>" .. "</code><ul><li>The <code>" .. "</code> block is found under TEXT.</li></ul></li></ul></li></ul></details>

In [15]:
#!response

category = 'day'
values = 'tip_percent'

#### Substep. &nbsp;&nbsp; Generate Box Plot

#!blhint

- Grab a freestyle block, type &nbsp;`x=category`
- Grab a freestyle block, type &nbsp;`y=values`
- `with px do box using` &nbsp;<-&nbsp; `create list with` &nbsp;:
    - variable `df`
    - `x=category`
    - `y=values`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li><li><code>A</code> &nbsp;&lt;-&nbsp; <code>B</code>&nbsp; means insert block B into the hole in block A.</li></ul></li><li>Blocks<ul><li>freestyle<ul><li>Unless specifically instructed, use the first block from the FREESTYLE menu.</li></ul></li><li><code>with .. do .. using ..</code><ul><li>The <code>with .. do .. using ..</code> block is found under VARIABLES.</li> <li>If <code>do</code> dropdown says "!Not populated until you execute code", click anywhere in the notebook tab, then click try "Run All Above Selected Cell" from the "Run" menu.</li> <li>If <code>with .. do .. using ..</code> block does not want to snap together nicely with the <code>set .. to</code> block, try dragging the <code>set .. to</code> block instead.</li></ul></li><li><code>create list with</code><ul><li>The <code>create list with</code> block is found under LISTS.</li> <li>You can use the <code>+ -</code> controls on the block to change the number of notches. Unless specifically instructed, the block should not have any empty notches when you click Blocks to Code. </li></ul></li><li>variable<ul><li>After it is created, each variable has its own block at the end of the VARIABLES menu.</li></ul></li></ul></li></ul></details>

In [16]:
#!response

px.box(df, x=category, y=values)

#!action

Look at the chart to answer the following questions.

#!action

Q1. Compare the 5 number summary for the box plots of each day of the week.

#!response

#!action

Q2. After seeing the box plots, do you think the tip percent depends on the day of the week when the party came?

#!response

#!action

Generate the box plot for Tip Percent vs Time.

#### Substep. &nbsp;&nbsp; Set Columns as Category and Value


#!blhint

- `set category to` &nbsp;:&nbsp; `" time "`
- `set values to` &nbsp;:&nbsp; `" tip_percent "`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li></ul></li><li>Blocks<ul><li><code>set .. to</code><ul><li>The <code>set .. to</code> block is found under VARIABLES.</li> <li>A variable must have been created before it will appear in the dropdown.</li></ul></li><li><code>" .. "</code><ul><li>The <code>" .. "</code> block is found under TEXT.</li></ul></li></ul></li></ul></details>

In [17]:
#!response

category = 'time'
values = 'tip_percent'

#### Substep. &nbsp;&nbsp; Generate Box Plot

#!blhint

- Grab a freestyle block, type &nbsp;`x=category`
- Grab a freestyle block, type &nbsp;`y=values`
- `with px do box using` &nbsp;<-&nbsp; `create list with` &nbsp;:
    - variable `df`
    - `x=category`
    - `y=values`

<details><summary>Blockly Hints</summary><ul><li>Notation<ul><li><code>A</code> &nbsp;:&nbsp; <code>B</code>&nbsp; means snap blocks A and B together.</li><li><code>A</code> &nbsp;&lt;-&nbsp; <code>B</code>&nbsp; means insert block B into the hole in block A.</li></ul></li><li>Blocks<ul><li>freestyle<ul><li>Unless specifically instructed, use the first block from the FREESTYLE menu.</li></ul></li><li><code>with .. do .. using ..</code><ul><li>The <code>with .. do .. using ..</code> block is found under VARIABLES.</li> <li>If <code>do</code> dropdown says "!Not populated until you execute code", click anywhere in the notebook tab, then click try "Run All Above Selected Cell" from the "Run" menu.</li> <li>If <code>with .. do .. using ..</code> block does not want to snap together nicely with the <code>set .. to</code> block, try dragging the <code>set .. to</code> block instead.</li></ul></li><li><code>create list with</code><ul><li>The <code>create list with</code> block is found under LISTS.</li> <li>You can use the <code>+ -</code> controls on the block to change the number of notches. Unless specifically instructed, the block should not have any empty notches when you click Blocks to Code. </li></ul></li><li>variable<ul><li>After it is created, each variable has its own block at the end of the VARIABLES menu.</li></ul></li></ul></li></ul></details>

In [18]:
#!response

px.box(df, x=category, y=values)

#!action

Look at the chart to answer the following questions.

#!action

Q1. Compare the 5 number summary for the Lunch box plot to the Dinner box plot.

#!response

#!action

Q2. After seeing the box plots, do you think the tip percent depends on the time of day when the party came?

#!response